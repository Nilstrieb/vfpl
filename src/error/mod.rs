mod display;
mod random;
mod span;

pub use display::display_error;
pub use random::random_ident;
pub use span::Span;

/// A trait for any error emitted during process
pub trait CompilerError {
    /// Return the span where the error occurred, for nicer error reporting. If no span is found,
    /// use [Span::dummy].
    fn span(&self) -> Span;

    /// The main message of the compiler error. It should be polite and not too long, but doesn't
    /// need extra verbosity.
    ///
    /// This needs to be a full sentence.
    ///
    /// # Examples
    ///
    /// `You have tried to add a String to a function.`
    fn message(&self) -> String;

    /// An additional note to give more context to the message. It can be long, but also needs to be
    /// polite.
    ///
    /// This needs to be a full sentence.
    ///
    /// # Examples
    /// `Due to the constraints of our spacetime, we have not found a sensible
    /// way to add a String to a function.`
    fn note(&self) -> Option<String> {
        None
    }

    /// A suggestion about what the programmer could do to resolve the error. It must be polite,
    /// and is allowed to be very wrong and not what the programmer intended, but it should fix the problem.
    ///
    /// This needs to be appendable to the phrase `You could `
    ///
    /// # Example
    /// `add another String to the String, for example "Function".`
    fn suggestion(&self) -> Option<String> {
        None
    }
}
