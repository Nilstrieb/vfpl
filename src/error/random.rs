pub fn random_ident() -> String {
    "double_value".to_string()
}
