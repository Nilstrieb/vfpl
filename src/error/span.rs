#![allow(dead_code)]

///
/// A Span points to a location in the source code
///
/// The start is inclusive, the end is exclusive
///
/// The special state, where both start and end are [usize::MAX] stands for the EOF span
#[derive(Debug, Copy, Clone, PartialEq, Eq, Default)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl Span {
    /// Create a span with a start and length
    pub fn start_len(start: usize, len: usize) -> Self {
        Self {
            start,
            end: start + len,
        }
    }

    /// Create a span with a start and end
    pub fn start_end(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    /// Create a span for a single character starting at `start`
    pub fn single(start: usize) -> Self {
        Self {
            start,
            end: start + 1,
        }
    }

    /// Create an empty span
    pub fn dummy() -> Self {
        Self { start: 0, end: 0 }
    }

    pub fn eof() -> Self {
        Self {
            start: usize::MAX,
            end: usize::MAX,
        }
    }

    /// Checks whether the Span is the dummy span
    pub fn is_dummy(&self) -> bool {
        *self == Span::dummy()
    }

    /// Extends the span by the second one
    /// The other one has to be after the current one
    pub fn extend(&self, other: Span) -> Span {
        if other.is_dummy() {
            return *self;
        }
        if self.is_dummy() {
            return other;
        }

        debug_assert!(self.start <= other.start);
        debug_assert!(self.end <= other.end);
        Span {
            start: self.start,
            end: other.end,
        }
    }

    /// Extends the span by the second one, if it exists
    /// The other one has to be after the current one, if it exists
    pub fn option_extend(&self, other: Option<Span>) -> Span {
        match other {
            None => *self,
            Some(span) => self.extend(span),
        }
    }

    pub fn len(&self) -> usize {
        self.end - self.start
    }
}
