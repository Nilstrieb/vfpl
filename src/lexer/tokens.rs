use crate::error::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub span: Span,
    pub kind: TokenType,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    Please,

    Ident(String),
    // Keywords
    As,
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    Initialize,
    Variable,
    With,
    The,
    Value,
    Of,
    Set,
    To,
    From,
    By,
    Take,
    End,
    Check,
    Whether,
    Then,
    Do,
    Otherwise,
    Break,
    Out,
    This,
    Create,
    Function,
    Parameter,
    Parameters,
    That,
    Returns,
    Call,
    No,
    Argument,
    Arguments,
    Go,
    Sleep,
    Absent,
    Null,
    NoValue,
    Undefined,
    True,
    False,
    Does,
    Has,
    Is,

    // literals
    String(String),
    Int(i64),
    Float(f64),

    // symbols
    /// .
    Dot,
    /// (
    ParenOpen,
    /// )
    ParenClose,
    /// ,
    Comma,
}

impl Display for TokenType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenType::Please => f.write_str("keyword `please`"),
            TokenType::Ident(name) => f.write_str(&format!("`{}`", name)),
            TokenType::As => f.write_str("keyword `as`"),
            TokenType::Add => f.write_str("keyword `add`"),
            TokenType::Subtract => f.write_str("keyword `subtract`"),
            TokenType::Multiply => f.write_str("keyword `multiply`"),
            TokenType::Divide => f.write_str("keyword `divide`"),
            TokenType::Modulo => f.write_str("keyword `modulo`"),
            TokenType::Initialize => f.write_str("keyword `initialize`"),
            TokenType::Variable => f.write_str("keyword `variable`"),
            TokenType::With => f.write_str("keyword `with`"),
            TokenType::The => f.write_str("keyword `the`"),
            TokenType::Value => f.write_str("keyword `value`"),
            TokenType::Of => f.write_str("keyword `of`"),
            TokenType::Set => f.write_str("keyword `set`"),
            TokenType::To => f.write_str("keyword `to`"),
            TokenType::From => f.write_str("keyword `from`"),
            TokenType::By => f.write_str("keyword `by`"),
            TokenType::Take => f.write_str("keyword `take`"),
            TokenType::End => f.write_str("keyword `end`"),
            TokenType::Check => f.write_str("keyword `check`"),
            TokenType::Whether => f.write_str("keyword `whether`"),
            TokenType::Then => f.write_str("keyword `then`"),
            TokenType::Do => f.write_str("keyword `do`"),
            TokenType::Otherwise => f.write_str("keyword `otherwise`"),
            TokenType::Break => f.write_str("keyword `break`"),
            TokenType::Out => f.write_str("keyword `out`"),
            TokenType::This => f.write_str("keyword `this`"),
            TokenType::Create => f.write_str("keyword `create`"),
            TokenType::Function => f.write_str("keyword `function`"),
            TokenType::Parameter => f.write_str("keyword `parameter`"),
            TokenType::Parameters => f.write_str("keyword `parameters`"),
            TokenType::That => f.write_str("keyword `that`"),
            TokenType::Returns => f.write_str("keyword `returns`"),
            TokenType::Go => f.write_str("keyword `go`"),
            TokenType::Sleep => f.write_str("keyword `sleep`"),
            TokenType::Absent => f.write_str("keyword `absent`"),
            TokenType::Null => f.write_str("keyword `null`"),
            TokenType::NoValue => f.write_str("keyword `no value` )"),
            TokenType::Undefined => f.write_str("keyword `undefined`"),
            TokenType::True => f.write_str("keyword `true`"),
            TokenType::False => f.write_str("keyword `false`"),
            TokenType::String(value) => f.write_str(&format!("`\"{}\"`", value)),
            TokenType::Int(value) => f.write_str(&format!("`{}`", value)),
            TokenType::Float(value) => f.write_str(&format!("`{}`", value)),
            TokenType::Dot => f.write_str("`.`"),
            TokenType::ParenOpen => f.write_str("`(`"),
            TokenType::ParenClose => f.write_str("`)`"),
            TokenType::Comma => f.write_str("`,`"),
        }
    }
}
