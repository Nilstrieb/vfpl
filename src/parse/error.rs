use crate::error::{CompilerError, Span};
use crate::VfplError;

#[derive(Debug)]
pub(super) struct ParseError {
    span: Span,
    message: String,
    note: Option<String>,
    suggestion: Option<String>,
}

impl ParseError {
    pub(super) fn simple(span: Span, message: String) -> Self {
        Self {
            span,
            message,
            note: None,
            suggestion: None,
        }
    }

    pub(super) fn full(span: Span, message: String, note: String, suggestion: String) -> Self {
        Self {
            span,
            message,
            note: Some(note),
            suggestion: Some(suggestion),
        }
    }
}

impl CompilerError for ParseError {
    fn span(&self) -> Span {
        self.span
    }

    fn message(&self) -> String {
        self.message.clone()
    }

    fn note(&self) -> Option<String> {
        self.note.clone()
    }

    fn suggestion(&self) -> Option<String> {
        self.suggestion.clone()
    }
}

impl From<ParseError> for VfplError {
    fn from(error: ParseError) -> Self {
        Self {
            span: error.span(),
            message: error.message(),
            note: error.note(),
            suggestion: error.suggestion(),
        }
    }
}
