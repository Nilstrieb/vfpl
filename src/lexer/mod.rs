use std::option::Option::Some;
use std::str::CharIndices;

use peekmore::{PeekMore, PeekMoreIterator};
use unicode_xid::UnicodeXID;

use tokens::Token;

use crate::error::{CompilerError, Span};
use crate::lexer::helper::compute_keyword;
use crate::lexer::tokens::TokenKind;

mod helper;
#[cfg(test)]
mod test;
pub mod tokens;

pub type LexerResult<T> = Result<T, LexerError>;

#[derive(Debug)]
pub struct Lexer<'a> {
    char_indices: PeekMoreIterator<CharIndices<'a>>,
}

impl Lexer<'_> {
    pub fn next(&mut self) -> Option<(usize, char)> {
        self.char_indices.next()
    }

    /// Consume elements n times
    pub fn consume_elements(&mut self, n: usize) {
        for _ in 0..n {
            self.next();
        }
    }

    pub fn peek(&mut self) -> Option<(usize, char)> {
        self.char_indices.peek().copied()
    }

    pub fn peek_nth(&mut self, n: usize) -> Option<(usize, char)> {
        self.char_indices.peek_nth(n).copied()
    }

    pub fn new(str: &str) -> Lexer {
        Lexer {
            char_indices: str.char_indices().peekmore(),
        }
    }

    pub fn compute_tokens(&mut self) -> LexerResult<Vec<Token>> {
        let mut tokens = Vec::new();
        while let Some((idx, char)) = self.next() {
            match char {
                '.' => tokens.push(Token::new_from_single(TokenKind::Dot, idx)),
                '(' => tokens.push(Token::new_from_single(TokenKind::ParenOpen, idx)),
                ')' => tokens.push(Token::new_from_single(TokenKind::ParenClose, idx)),
                ',' => tokens.push(Token::new_from_single(TokenKind::Comma, idx)),
                '\"' => tokens.push(self.compute_string(idx)?),
                '<' => {
                    if self.peek_nth(0).ok_or(LexerError::UnexpectedEOF)?.1 == '!'
                        && self.peek_nth(1).ok_or(LexerError::UnexpectedEOF)?.1 == '-'
                        && self.peek_nth(2).ok_or(LexerError::UnexpectedEOF)?.1 == '-'
                    {
                        while let Some((_, char)) = self.next() {
                            if char == '-' {
                                if self.peek_nth(0).ok_or(LexerError::UnexpectedEOF)?.1 == '-'
                                    && self.peek_nth(1).ok_or(LexerError::UnexpectedEOF)?.1 == '>'
                                {
                                    self.consume_elements(2);
                                    break;
                                }
                            } else {
                                continue;
                            }
                        }
                    } else {
                        return Err(LexerError::InvalidCharacter(Span::single(idx)));
                    }
                }
                '0'..='9' | '-' => tokens.push(self.compute_number(char, idx)?),
                other if other.is_whitespace() => continue,
                other if other.is_xid_start() => tokens.push(self.compute_identifier(char, idx)),
                _ => return Err(LexerError::InvalidCharacter(Span::single(idx))),
            }
        }

        tokens.push(Token::eof());
        Ok(tokens)
    }

    fn compute_identifier(&mut self, char: char, idx: usize) -> Token {
        let mut identifier = String::from(char);
        while let Some((_, char)) = self.peek() {
            if char.is_xid_continue() {
                self.next();
                identifier.push(char);
            } else {
                break;
            }
        }
        identifier = identifier.to_lowercase();
        let end = identifier.len();
        let kind = compute_keyword(&identifier).unwrap_or(TokenKind::Ident(identifier));
        Token::new_from_len(kind, idx, end)
    }

    fn compute_number(&mut self, char: char, idx: usize) -> LexerResult<Token> {
        let mut number = String::from(char);
        let mut end = loop {
            if let Some((i, char)) = self.peek() {
                match char {
                    '0'..='9' => {
                        self.next();
                        number.push(char);
                    }
                    other if other.is_xid_start() => {
                        return Err(LexerError::InvalidCharacter(Span::single(i)));
                    }
                    _ => {
                        // This works because the character before it will be 1 byte long (0..9 || -)
                        break i;
                    }
                }
            } else {
                break idx + number.len();
            }
        };

        if let Some((i, char)) = self.peek() {
            if char == '.' {
                if let Some((i, char)) = self.peek_nth(1) {
                    println!("{}", char);
                    if char.is_ascii_digit() {
                        number.push('.');
                        self.next();
                        number.push(char);
                        self.next();

                        end = loop {
                            if let Some((i, char)) = self.peek() {
                                match char {
                                    '0'..='9' => {
                                        self.next();
                                        number.push(char)
                                    }
                                    other if other.is_xid_start() => {
                                        return Err(LexerError::InvalidCharacter(Span::single(i)));
                                    }
                                    _ => {
                                        // This works because the character before it will be 1 byte long (0..9 || -)
                                        break i;
                                    }
                                }
                            } else {
                                break idx + number.len();
                            }
                        }
                    } else if char.is_xid_start() {
                        return Err(LexerError::InvalidCharacter(Span::single(i)));
                    }
                }
            } else if char.is_ascii_alphabetic() {
                return Err(LexerError::LetterInNumber(Span::start_end(idx, i)));
            }
        }

        if number.contains('.') {
            let number = number
                .parse::<f64>()
                .map_err(|_| LexerError::FloatParseError(Span::start_end(idx, end)))?;
            Ok(Token::new(TokenKind::Float(number), idx, end))
        } else {
            let number = number
                .parse::<i64>()
                .map_err(|_| LexerError::IntParseError(Span::start_end(idx, end)))?;
            Ok(Token::new(TokenKind::Int(number), idx, end))
        }
    }

    fn compute_string(&mut self, start: usize) -> LexerResult<Token> {
        let mut str = String::new();
        while let Some((idx, char)) = self.next() {
            match char {
                '\\' => {
                    let (_, next_char) = self.next().ok_or(LexerError::UnexpectedEOF)?;
                    match next_char {
                        '"' => str.push('\"'),
                        'n' => str.push('\n'),
                        'r' => str.push('\r'),
                        't' => str.push('\t'),
                        '0' => str.push('\0'),
                        '\\' => str.push('\\'),
                        _ => {
                            return Err(LexerError::InvalidEscapeCharacter(Span::start_end(
                                idx,
                                idx + 1,
                            )));
                        }
                    }
                }
                '"' => {
                    return Ok(Token::new(TokenKind::String(str), start, idx));
                }
                other => str.push(other),
            }
        }
        Err(LexerError::UnexpectedEOF)
    }
}

#[derive(Debug)]
pub enum LexerError {
    UnexpectedEOF,
    IntParseError(Span),
    FloatParseError(Span),
    InvalidCharacter(Span),
    InvalidEscapeCharacter(Span),
    LetterInNumber(Span),
}

impl CompilerError for LexerError {
    fn span(&self) -> Span {
        match self {
            LexerError::UnexpectedEOF => Span::dummy(),
            LexerError::IntParseError(span) => *span,
            LexerError::FloatParseError(span) => *span,
            LexerError::InvalidCharacter(span) => *span,
            LexerError::InvalidEscapeCharacter(span) => *span,
            LexerError::LetterInNumber(span) => *span,
        }
    }

    fn message(&self) -> String {
        match self {
            LexerError::UnexpectedEOF => "Unexpected EOF".to_string(),
            LexerError::IntParseError(_) => "Error parsing integer literal".to_string(),
            LexerError::FloatParseError(_) => "Error parsing float literal".to_string(),
            LexerError::InvalidCharacter(_) => "Invalid character".to_string(),
            LexerError::InvalidEscapeCharacter(_) => "Invalid escape character".to_string(),
            LexerError::LetterInNumber(_) => "Invalid letter in number".to_string(),
        }
    }

    fn note(&self) -> Option<String> {
        None
    }
}
