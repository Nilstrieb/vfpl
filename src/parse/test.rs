//// helpers

use crate::error::Span;
use crate::lexer::tokens::{
    Token,
    TokenKind::{self, *},
};
use crate::parse::Parser;

fn token(kind: TokenKind) -> Token {
    Token {
        span: Span::dummy(),
        kind,
    }
}

fn parse<T, F, R>(tokens: T, parse_rule_fn: F) -> R
where
    T: Into<Vec<Token>>,
    F: FnOnce(&mut Parser) -> R,
{
    let vec = tokens.into();
    let mut parser = Parser::new(vec.into_iter());
    parse_rule_fn(&mut parser)
}

//// tests

#[test]
fn it_works() {
    let snap = "test";
    insta::assert_debug_snapshot!(snap);
}

#[test]
fn single_string_literal() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::literal);
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn different_literals() {
    let literals = [
        [String("test".to_string())].map(token),
        [Int(425)].map(token),
        [Float(2465.67)].map(token),
        [Absent].map(token),
        [NoValue].map(token),
        [Null].map(token),
        [Undefined].map(token),
        [True].map(token),
        [False].map(token),
    ];

    let parsed = literals.map(|tokens| parse(tokens, Parser::literal));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn ident_ty() {
    let tokens = [Ident("Test".to_string())].map(token);
    let parsed = parse(tokens, Parser::ty);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn nullable_ty() {
    let tys = [
        [Absent].map(token),
        [NoValue].map(token),
        [Null].map(token),
        [Undefined].map(token),
    ];
    let parsed = tys.map(|tokens| parse(tokens, Parser::ty));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn other_tys() {
    let tys = [
        [Ident("Boolean".to_string())].map(token),
        [Ident("Integer".to_string())].map(token),
        [Ident("String".to_string())].map(token),
        [Ident("Float".to_string())].map(token),
    ];
    let parsed = tys.map(|tokens| parse(tokens, Parser::ty));

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn literal_call_expr() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::call_expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn literal_expr() {
    let tokens = [String("string".to_string())].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn ident_literal() {
    let tokens = [Ident("uwu".to_string())].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn not_equal_expr() {
    let tokens = [Absent, Does, Not, Have, The, Value, Null].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn equal_expr() {
    let tokens = [Absent, Has, The, Value, Absent].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn greater_than_expr() {
    let tokens = [Absent, Is, Greater, Than, Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn less_than_expr() {
    let tokens = [Absent, Is, Less, Than, Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn greater_equal_than_expr() {
    let tokens = [Absent, Is, Greater, Or, Equal, Than, Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn less_equal_than_expr() {
    let tokens = [Absent, Is, Less, Or, Equal, Than, Undefined].map(token);
    let parsed = parse(tokens, Parser::expr);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn typed_ident_absent() {
    let tokens = [Ident("name".to_string()), As, Absent].map(token);
    let parsed = parse(tokens, Parser::typed_ident);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn init_variable_string() {
    let tokens = [
        Initialize,
        Variable,
        Ident("name".to_string()),
        As,
        Ident("string".to_string()),
        With,
        The,
        Value,
        Of,
        String("Ferris".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::var_init);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn set_variable_string() {
    let tokens = [
        Set,
        The,
        Variable,
        Ident("name".to_string()),
        To,
        The,
        Value,
        Of,
        String("Ferris".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::var_set);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn add_number() {
    let tokens = [Add, Int(0), To, Ident("counter".to_string())].map(token);
    let parsed = parse(tokens, Parser::add);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn subtract_number() {
    let tokens = [Sub, Int(1), From, Ident("counter".to_string())].map(token);
    let parsed = parse(tokens, Parser::subtract);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn divide_number() {
    let tokens = [Div, Ident("counter".to_string()), By, Int(2)].map(token);
    let parsed = parse(tokens, Parser::divide);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn multiply_number() {
    let tokens = [Mul, Ident("counter".to_string()), With, Int(3)].map(token);
    let parsed = parse(tokens, Parser::multiply);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn mod_number() {
    let tokens = [Take, Ident("counter".to_string()), Mod, Int(4)].map(token);
    let parsed = parse(tokens, Parser::modulo);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn terminate() {
    let tokens = [Go, To, Sleep].map(token);
    let parsed = parse(tokens, Parser::terminate);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_no_args() {
    let tokens = [Call, Ident("run".to_string()), With, No, Arguments].map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_single_arg() {
    let tokens = [
        Call,
        Ident("print".to_string()),
        With,
        The,
        Argument,
        Int(0),
        As,
        Ident("printable".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_two_args() {
    let tokens = [
        Call,
        Ident("add".to_string()),
        With,
        The,
        Arguments,
        Int(2),
        As,
        Ident("a".to_string()),
        And,
        Int(3),
        As,
        Ident("b".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn call_three_args() {
    let tokens = [
        Call,
        Ident("ternary".to_string()),
        With,
        The,
        Arguments,
        True,
        As,
        Ident("cond".to_string()),
        Comma,
        Int(0),
        As,
        Ident("then".to_string()),
        And,
        Int(3),
        As,
        Ident("else".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::call);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn add_stmt() {
    let tokens = [Please, Add, Int(0), To, Ident("A".to_string()), Dot].map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_empty_body_true_cond() {
    let tokens = [Check, Whether, True, Comma, Then, Do, Please, End, Check].map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_empty_body_complex_cond() {
    let tokens = [
        Check, Whether, True, Has, The, Value, Absent, Comma, Then, Do, Please, End, Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_single_stmt_body() {
    let tokens = [
        Check, Whether, True, Has, The, Value, Absent, Comma, Then, Do, Please, Go, To, Sleep, Dot,
        Please, End, Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn if_multi_stmt_body() {
    let tokens = [
        Check,
        Whether,
        True,
        Has,
        The,
        Value,
        Absent,
        Comma,
        Then,
        Do,
        Please,
        Go,
        To,
        Sleep,
        Dot,
        Please,
        Add,
        Int(5),
        To,
        Ident("A".to_string()),
        Dot,
        Please,
        End,
        Check,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::if_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_number() {
    let tokens = [Return, Int(9), From, The, Function].map(token);
    let parsed = parse(tokens, Parser::return_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_stmt() {
    let tokens = [Please, Return, Int(9), From, The, Function, Dot].map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn r#break() {
    let tokens = [Break, Out, Of, This, While].map(token);
    let parsed = parse(tokens, Parser::break_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn break_stmt() {
    let tokens = [Please, Break, Out, Of, This, While, Dot].map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn while_loop_no_body() {
    let tokens = [Repeat, While, True, Do, Please, End, While].map(token);
    let parsed = parse(tokens, Parser::while_stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn while_stmt_with_body() {
    let tokens = [
        Please,
        Repeat,
        While,
        Absent,
        Has,
        The,
        Value,
        False,
        Do,
        Please,
        Add,
        Int(9),
        To,
        Ident("A".to_string()),
        Dot,
        Please,
        End,
        While,
        Dot,
    ]
    .map(token);
    let parsed = parse(tokens, Parser::stmt);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn return_ty_absent() {
    let tokens = [That, Returns, Absent].map(token);
    let parsed = parse(tokens, Parser::fn_return);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_no_params_empty_body() {
    let tokens = [
        Create,
        Function,
        Ident("void".to_string()),
        With,
        No,
        Parameters,
        That,
        Returns,
        Absent,
        Please,
        End,
        Function,
        Ident("void".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_single_param_emtpy_body() {
    let tokens = [
        Create,
        Function,
        Ident("print".to_string()),
        With,
        The,
        Parameter,
        Ident("_".to_string()),
        As,
        Absent,
        That,
        Returns,
        Absent,
        Please,
        End,
        Function,
        Ident("print".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_two_param_emtpy_body() {
    let tokens = [
        Create,
        Function,
        Ident("add".to_string()),
        With,
        The,
        Parameters,
        Ident("_".to_string()),
        As,
        Absent,
        And,
        Ident("_hi".to_string()),
        As,
        Null,
        That,
        Returns,
        Absent,
        Please,
        End,
        Function,
        Ident("add".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn fn_decl_three_param_with_body() {
    let tokens = [
        Create,
        Function,
        Ident("add".to_string()),
        With,
        The,
        Parameters,
        Ident("_".to_string()),
        As,
        Absent,
        Comma,
        Ident("a".to_string()),
        As,
        NoValue,
        And,
        Ident("_hi".to_string()),
        As,
        Null,
        That,
        Returns,
        Absent,
        Please,
        Add,
        Int(5),
        To,
        Ident("a".to_string()),
        Dot,
        Please,
        End,
        Function,
        Ident("add".to_string()),
    ]
    .map(token);
    let parsed = parse(tokens, Parser::fn_decl);

    insta::assert_debug_snapshot!(parsed);
}
