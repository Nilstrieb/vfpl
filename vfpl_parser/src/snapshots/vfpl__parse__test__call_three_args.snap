---
source: src/parse/test.rs
expression: parsed

---
Ok(
    Call {
        span: Span {
            start: 0,
            end: 0,
        },
        fn_name: "ternary",
        args: CallArgs {
            span: Span {
                start: 0,
                end: 0,
            },
            args: [
                CallArg {
                    span: Span {
                        start: 0,
                        end: 0,
                    },
                    expr: Literal(
                        Literal {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            kind: True,
                        },
                    ),
                    name: "cond",
                },
                CallArg {
                    span: Span {
                        start: 0,
                        end: 0,
                    },
                    expr: Literal(
                        Literal {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            kind: Int(
                                0,
                            ),
                        },
                    ),
                    name: "then",
                },
                CallArg {
                    span: Span {
                        start: 0,
                        end: 0,
                    },
                    expr: Literal(
                        Literal {
                            span: Span {
                                start: 0,
                                end: 0,
                            },
                            kind: Int(
                                3,
                            ),
                        },
                    ),
                    name: "else",
                },
            ],
        },
    },
)
